cmake/external/ngraph.cmake

```diff
@@ -0,0 +1,57 @@
+# Copyright(C) 2019 Intel Corporation
+# Licensed under the MIT License
+
+include (ExternalProject)
+
+set(ngraph_ROOT_DIR ${CMAKE_CURRENT_BINARY_DIR}/external/ngraph)
+set(ngraph_INSTALL_DIR ${ngraph_ROOT_DIR})
+set(ngraph_INCLUDE_DIRS ${ngraph_INSTALL_DIR}/include)
+set(ngraph_LIBRARIES ${ngraph_INSTALL_DIR}/lib)
+set(ngraph_SRC ${CMAKE_CURRENT_BINARY_DIR}/ngraph/src/project_ngraph)
+set(prebuilt_ONNX_SOURCE_DIR "${PROJECT_SOURCE_DIR}/external/onnx")
+set(prebuilt_ONNX_BINARY_DIR "${CMAKE_CURRENT_BINARY_DIR}/onnx")
+set(ngraph_URL "https://github.com/NervanaSystems/ngraph.git")
+set(ngraph_TAG "v0.18.0")
+
+# Libraries for python package.
+set(NGRAPH_SHARED_LIB libngraph.so)
+set(NGRAPH_CODEGEN_SHARED_LIB libcodegen.so)
+set(NGRAPH_CPU_BACKEND_SHARED_LIB libcpu_backend.so)
+set(NGRAPH_IOMP5MD_SHARED_LIB libiomp5.so)
+set(NGRAPH_MKLDNN_SHARED_LIB libmkldnn.so)
+set(NGRAPH_MKLML_SHARED_LIB libmklml_intel.so)
+if("${CMAKE_BUILD_TYPE}" STREQUAL "Debug")
+    set(NGRAPH_TBB_SHARED_LIB libtbb_debug.so)
+    set(NGRAPH_TBB_SHARED_LIB_2 libtbb_debug.so.2)
+else()
+    set(NGRAPH_TBB_SHARED_LIB libtbb.so)
+    set(NGRAPH_TBB_SHARED_LIB_2 libtbb.so.2)
+endif()
+
+ExternalProject_Add(project_ngraph
```

Comment:
Personally, I prefer to add this as a submodule. It's better to have no network downloading during the build. In the last Friday and weekend, I was fighting with a strange build issue. Some of our CI builds can't download eigen. Eigen is added as an external project like this, not a submodule. If it is a submodule, it's will be downloaded by VSTS, not cmake. It will less likely have a download issue. @jywu-msft , what do you think?

Reply:
I'm just following mkl-dnn ep style here, which also does it through external_project. 
Also in the current state,  we are patching couple of files after clone. But, if we want switch to submodule style, I'll have to see how if this can be done.

Reply:
I kind of prefer cmake external projects actually. it's the first time I heard about download issues, which I am guessing should be rare.
it's easier to read all the build options, patching, version dependency etc. needed for the external project if it's in its own cmake file. 
for ngraph, we are doing a patch command, which is more easily supported with cmake. 
in my experience git submodule is more prone to making mistakes during development. if we treat the external repo as read-only (not actively developed with the parent repo), it seems easier to just do git external project.

Reply:
This is the original discussion: https://github.com/Microsoft/onnxruntime/issues/662#issuecomment-474840133

I personally agree with him on this point.

Reply:
Yes, perhaps but that is a separate discussion whether we should build external dependencies from source or they should be pre-installed. (that works well with docker images)

Whether we use git submodule or cmake external project, we are building from the source of a different repository. in both cases if there's a network issue, wouldn't that prevent us from pulling the sources in either case?

Reply:
First, VSTS handle network errors better than cmake. some network errors are temporary and it can be retried.
Second, as he said, internet network in China is very bad. If it is a submodule, it only need be download once. But if it is in build dir, whenever the  build dir is cleaned(which is often needed), he need to download it again.
As this is a broader discussion, let me close this conversation now. We can discuss it later.